internal/user/user_profile.go

```diff
@@ -1,38 +1,41 @@
 package user
 
 import (
+	"github.com/kopia/kopia/internal/crypto"
 	"github.com/kopia/kopia/repo/manifest"
 )
 
 // Profile describes information about a single user.
 type Profile struct {
 	ManifestID manifest.ID `json:"-"`
 
-	Username            string `json:"username"`
-	PasswordHashVersion int    `json:"passwordHashVersion"` // indicates how password is hashed
-	PasswordHash        []byte `json:"passwordHash"`
+	Username               string `json:"username"`
+	PasswordHashVersion    int    `json:"passwordHashVersion,omitempty"` // indicates how password is hashed, deprecated in favor of KeyDerivationAlgorithm
+	KeyDerivationAlgorithm string `json:"keyDerivationAlgorithm,omitempty"`
+	PasswordHash           []byte `json:"passwordHash"`
 }
 
 // SetPassword changes the password for a user profile.
-func (p *Profile) SetPassword(password, keyDerivationAlgorithm string) error {
-	return p.setPasswordV1(password, keyDerivationAlgorithm)
+func (p *Profile) SetPassword(password string) error {
+	return p.setPassword(password)
 }
 
 // IsValidPassword determines whether the password is valid for a given user.
-func (p *Profile) IsValidPassword(password, keyDerivationAlgorithm string) bool {
+func (p *Profile) IsValidPassword(password string) bool {
 	if p == nil {
 		// if the user is invalid, return false but use the same amount of time as when we
 		// compare against valid user to avoid revealing whether the user account exists.
-		isValidPasswordV1(password, dummyV1HashThatNeverMatchesAnyPassword, keyDerivationAlgorithm)
+		isValidPassword(password, dummyV1HashThatNeverMatchesAnyPassword, crypto.DefaultKeyDerivationAlgorithm)
```

Comment:
Well we are kind of out of options here... there is going to be some change in processing time because for an attacker "default algorithm" time means a 50% chance of being a valid user and non-default would mean a valid user ..., but better than no processing.

Ideally, the algorithm for the `nil` case should be picked up randomly to protect against timing attacks, and then you'll need to make sure that randomization is done each time (even if `p != nil`).

Reply:
WDYT ?